packages/core/http/core-http-server-internal/src/csp/csp_directives.ts

```diff
@@ -83,60 +111,87 @@ export class CspDirectives {
       [...additionalValues, ...directiveValues].forEach((value) => {
         cspDirectives.addDirectiveValue(directiveName, value);
       });
-    });
+    }
+
+    for (const [directiveName, directiveValues] of reportOnlyDirectives.entries()) {
+      directiveValues.forEach((value) => {
+        cspDirectives.addDirectiveValue(directiveName, value, false);
+      });
+    }
 
     return cspDirectives;
   }
 }
 
-const parseConfigDirectives = (cspConfig: CspConfigType): Map<CspDirectiveName, string[]> => {
-  const map = new Map<CspDirectiveName, string[]>();
+const parseConfigDirectives = (cspConfig: CspConfigType): CspConfigDirectives => {
+  const enforceDirectives = new Map<CspDirectiveName, string[]>();
+  const reportOnlyDirectives = new Map<CspDirectiveName, string[]>();
 
   if (cspConfig.script_src?.length) {
-    map.set('script-src', cspConfig.script_src);
+    enforceDirectives.set('script-src', cspConfig.script_src);
   }
   if (cspConfig.disableUnsafeEval !== true) {
-    map.set('script-src', ["'unsafe-eval'", ...(map.get('script-src') ?? [])]);
+    enforceDirectives.set('script-src', [
+      "'unsafe-eval'",
+      ...(enforceDirectives.get('script-src') ?? []),
+    ]);
   }
   if (cspConfig.worker_src?.length) {
-    map.set('worker-src', cspConfig.worker_src);
+    enforceDirectives.set('worker-src', cspConfig.worker_src);
   }
   if (cspConfig.style_src?.length) {
-    map.set('style-src', cspConfig.style_src);
+    enforceDirectives.set('style-src', cspConfig.style_src);
   }
   if (cspConfig.connect_src?.length) {
-    map.set('connect-src', cspConfig.connect_src);
+    enforceDirectives.set('connect-src', cspConfig.connect_src);
   }
   if (cspConfig.default_src?.length) {
-    map.set('default-src', cspConfig.default_src);
+    enforceDirectives.set('default-src', cspConfig.default_src);
   }
   if (cspConfig.font_src?.length) {
-    map.set('font-src', cspConfig.font_src);
+    enforceDirectives.set('font-src', cspConfig.font_src);
   }
   if (cspConfig.frame_src?.length) {
-    map.set('frame-src', cspConfig.frame_src);
+    enforceDirectives.set('frame-src', cspConfig.frame_src);
   }
   if (cspConfig.img_src?.length) {
-    map.set('img-src', cspConfig.img_src);
+    enforceDirectives.set('img-src', cspConfig.img_src);
   }
   if (cspConfig.frame_ancestors?.length) {
-    map.set('frame-ancestors', cspConfig.frame_ancestors);
+    enforceDirectives.set('frame-ancestors', cspConfig.frame_ancestors);
   }
   if (cspConfig.report_uri?.length) {
-    map.set('report-uri', cspConfig.report_uri);
+    enforceDirectives.set('report-uri', cspConfig.report_uri);
+    reportOnlyDirectives.set('report-uri', cspConfig.report_uri);
   }
   if (cspConfig.report_to?.length) {
-    map.set('report-to', cspConfig.report_to);
+    enforceDirectives.set('report-to', cspConfig.report_to);
+    reportOnlyDirectives.set('report-to', cspConfig.report_to);
+  }
+
+  if (cspConfig.report_only?.form_action?.length) {
```

Comment:
__note__ I was about to suggest not adding report-only directives and header if neither `report_uri` nor `report_to` is set, as it'd be useless in many cases (we explicitly set `report_to` only in ESS right now). However, it feels like it would require quite a bit of conditional logic in all places where we return this header (`if CspConfig.DEFAULT.reportOnlyHeader != null then add header, otherwise - don't`), which sounds too expensive for the value we might get. So I'm totally fine with keeping it as is, unless you want to add this conditional logic :slightly_smiling_face:

Reply:
Right now the behavior is the same as I see from code for the enforce header as well. If neither `report_uri` nor `report_to` is provided we still set the `Content-Security-Policy` header. 

So if we consider adding conditional logic, it would be better to adjust it for both headers for consistency and it will indeed require a lot of if statements, therefore test updates. I agree that the profit of it might be probably negligible.

Reply:
> So if we consider adding conditional logic, it would be better to adjust it for both headers for consistency and it will indeed require a lot of if statements, therefore test updates. 

We don't really need this conditional logic for enforcing header because it, well, enforces CSP restrictions even if it cannot report them (e.g., `report-to` is set). But anyway, keeping it as is sounds good to me too 👍